kmath-core/src/commonMain/kotlin/kscience/kmath/expressions/SimpleAutoDiff.kt

```diff
@@ -0,0 +1,325 @@
+package kscience.kmath.expressions
+
+import kscience.kmath.linear.Point
+import kscience.kmath.operations.*
+import kscience.kmath.structures.asBuffer
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
+
+/*
+ * Implementation of backward-mode automatic differentiation.
+ * Initial gist by Roman Elizarov: https://gist.github.com/elizarov/1ad3a8583e88cb6ea7a0ad09bb591d3d
+ */
+
+
+public open class AutoDiffValue<out T>(public val value: T)
```

Comment:
I am sure that it may be replaced with sealed class to prevent one to extend AutoDiffValue with irrelevant object.

Reply:
The idea is that it could be extended anytime. Here `AutoDiffValue` is just a marker interface. It is possible to even replace it with an inline class, but we need performance measurements to make that change.